src/main/java/org/spongepowered/api/event/lifecycle/RegisterRegistryValueEvent.java

```diff
@@ -27,23 +27,93 @@
 import org.spongepowered.api.Engine;
 import org.spongepowered.api.ResourceKey;
 import org.spongepowered.api.event.GenericEvent;
-import org.spongepowered.api.registry.DefaultedRegistryValue;
+import org.spongepowered.api.registry.Registry;
+import org.spongepowered.api.registry.RegistryHolder;
+import org.spongepowered.api.registry.RegistryRegistrationSet;
 import org.spongepowered.api.registry.RegistryType;
 import org.spongepowered.eventgen.annotations.NoFactoryMethod;
 
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+
+/**
+ * Lifecycle event for registering entries to a {@link Registry}.
+ *
+ * <p>This event may be called multiple times for the same layer
+ * as the registry is being appended and as dependencies become
+ * available. Plugins should only execute code inside the consumers.</p>
+ *
+ * <p><strong>Note:</strong> Layers might be reloadable!</p>
+ */
 @NoFactoryMethod
 public interface RegisterRegistryValueEvent extends LifecycleEvent {
 
-    <T> RegistryStep<T> registry(RegistryType<T> registryType);
+    /**
+     * <p>Register new entries against a specific {@link Registry}.</p>
+     *
+     * <p>The consumer will be called if it matches the current set of
+     * registries being created.</p>
+     *
+     * @param registryType The registry type to append.
+     * @param consumer The consumer to be called if found.
+     */
+    default <T> void registry(RegistryType<T> registryType, Consumer<RegistryStep<T>> consumer) {
+        this.registry(registryType, (h, r) -> consumer.accept(r));
+    }
 
-    interface RegistryStep<T> {
+    /**
+     * <p>Register new entries against a specific {@link Registry}.</p>
+     *
+     * <p>The consumer will be called if it matches the current set of
+     * registries being created.</p>
+     *
+     * @param registryType The registry type to append.
+     * @param consumer The consumer to be called if found.
+     */
+    <T> void registry(RegistryType<T> registryType, BiConsumer<RegistryHolder, RegistryStep<T>> consumer);
 
-        RegistryStep<T> register(ResourceKey key, T value);
+    /**
+     * <p>Register new entries against a specific {@link Registry}.</p>
+     *
+     * <p>The consumer will be called if it matches the current set of
+     * registries being created.</p>
+     *
+     * @param registryType The registry type to append.
+     * @param consumer The consumer to be called if found.
+     * @param dependencies The list of registries to create before
+     *                     invoking the consumer.
+     */
+    <T> void registry(RegistryType<T> registryType, BiConsumer<RegistryHolder, RegistryStep<T>> consumer, final RegistryType<?>... dependencies);
```

Comment:
Can you provide reasoning in the documentation as to why registry dependencies would be useful? Also, if I am registering 3 registries with B and C dependent on A, how do I have their types at this point?